Source/ACE/Entity/Vendor.cs

```diff
@@ -159,38 +179,75 @@ public void BuyItemsStartTransaction(ObjectGuid vendorid, List<ItemProfile> item
+        /// Items have been purchased from vendor by player but we need to validate it.
         /// </summary>
         /// <param name="player"></param>
         /// <param name="items"></param>
-        public void BuyItemsCompleteTransaction(Player player, List<WorldObject> items)
+        public void BuyItemsValidateTransaction(Player player, List<WorldObject> items)
         {
-            // todo: remove unique items to vendor unique list.
-            // they where sold to a player.
+            // todo: add logic for temp stock items.
+            // remove items, if transaction fails then move temp stock items back into temp stock
+            player.HandleActionBuyFinalTransaction(this, items, true);
+        }
 
+        /// <summary>
+        /// Handles the final phase of the transaction.. removing unique items and updating players local
+        /// from vendors items list.
+        /// </summary>
+        /// <param name="player"></param>
+        /// <param name="items"></param>
+        public void BuyItemsFinalTransaction(Player player, List<WorldObject> items, bool valid)
+        {
             List<WorldObject> vendorlist = new List<WorldObject>();
+
             foreach (KeyValuePair<ObjectGuid, WorldObject> wo in defaultItemsForSale)
             {
                 vendorlist.Add(wo.Value);
             }
 
-            // todo: send more then default items.
-            player.TrackInteractiveObjects(vendorlist);
+            if (valid)
+            {
+                // todo: remove unique temp stock items.
+            }
+            else
+            {
+                // todo: re-add unique temp stock items.
+            }
             player.HandleActionApproachVendor(this, vendorlist);
         }
-    #endregion
+        #endregion
 
-    #region Sell Transactions
+        #region Sell Transactions
         public void SellItemsStartTransaction(List<WorldObject> items, Player player)
         {
-            // todo: calculate payment based on vendor multipliers
-            uint coin = 0;
-            foreach (WorldObject item in items)
+            // todo: filter out items vendor does not purchase.
+            player.HandleActionSellStartTransaction(items, Guid);
+        }
+
+        public void SellItemsValidateTransaction(Player player, List<WorldObject> items)
+        {
+            // todo: calculate payment based on multipliers correctly
+            uint payout = 0;
+            foreach (WorldObject wo in items)
             {
-                coin += (uint)item.Value;
+                if (wo.StackSize.HasValue && wo.StackSize.Value != 0)
+                    payout = payout + ((uint)wo.Value * wo.StackSize.Value);
+                else
+                    payout = payout + (uint)wo.Value;
             }
-            player.HandleActionSellItemsStartTransaction(items, Guid, coin);
+            player.HandleActionSellFinalTransaction(this, items, true, payout);
```

Comment:
Vendors have a Double Property that dictates what factor they buy at (PropertyDouble.BuyPrice / 37).  Payout should be scaled by that factor (ie, 0.8 would be typical, so you get 80% of value).